15th sprint was added:

AntonSik opened the pull request:
- packages
- playerRepository interface and his implementation
- removed activity-files into ui package

AntonSik committed:
15th sprint was added:
- packages
- playerRepository interface and his implementation
- removed activity-files into ui package

IvanGusevE commented on app/src/main/java/com/example/playlistmaker/domain/repository/PlayerRepository.kt:
```diff
@@ -0,0 +1,17 @@
+package com.example.playlistmaker.domain.repository
+
+import android.widget.ImageButton
+import android.widget.TextView
```
**_Критическая_**

Слой бизнес логики (domain-layer) не должен иметь зависимости presentation слоя. Чтобы получить необходимые состояния медиаплеера из data слоя, тебе нужно через репозиторий передать колбэки. Например:

**domain**
```
interface PlayerRepository {

     fun setOnPreparedCallback(callback: () -> Unit)
     
     fun setOnCompletionCallback(callback: () -> Unit)
     
     fun preparePlayer(recordsUrl: String?)
}
```

**data**
```
class PlayerRepositoryImpl(...) : PlayerRepository {
     
     private val preparedCallback: (() -> Unit)? = null

     override fun setOnPreparedCallback(callback: () -> Unit){
          preparedCallback = callback
     }
     
     fun setOnCompletionCallback(callback: () -> Unit){
         ...
     }
     
     fun preparePlayer(recordsUrl: String?){
        mediaPlayer.setDataSource(recordsUrl)
        mediaPlayer.prepareAsync()
        mediaPlayer.setOnPreparedListener {
            preparedCallback?.invoke()
        }
     }
}
```

**presentation**
```
playerRepository.setOnPreparedCallback{
     playBtn.isEnabled = true
}
```

IvanGusevE commented on app/src/main/java/com/example/playlistmaker/data/repository/PlayerRepositoryImpl.kt:
```diff
@@ -0,0 +1,91 @@
+package com.example.playlistmaker.data.repository
+
+import android.content.Context
+import android.media.MediaPlayer
+import android.os.Handler
+import android.os.Looper
+import android.widget.ImageButton
+import android.widget.TextView
```
**_Критическая_**
Слой данных (data-layer) тоже не должен знать о presentation слое.

IvanGusevE commented on app/src/main/java/com/example/playlistmaker/data/repository/PlayerRepositoryImpl.kt:
```diff
@@ -0,0 +1,91 @@
+
+class PlayerRepositoryImpl(private val context: Context) : PlayerRepository {
+
+    companion object {
+        private const val STATE_DEFAULT = 0
+        private const val STATE_PREPARED = 1
+        private const val STATE_PLAYING = 2
+        private const val STATE_PAUSED = 3
+        private const val UPDATING_DELAY = 400L
+    }
```
Есть предложение перенести состояния медиаплеера в domain слой, как enum class. Это позволит тебе добавить в интерфейс репозитория метод `setOnChangePlayerListener(listener: (State) -> Unit)`. Через него экран AudioPlayerActivity сможет получать текущее состояние mediaplayer из data слоя (реализация будет точно такая же, как и для prepared события).

IvanGusevE commented on app/src/main/java/com/example/playlistmaker/presentation/ui/AudioPlayerActivity.kt:
```diff
@@ -13,19 +10,16 @@ import android.widget.TextView
-        private const val UPDATING_DELAY = 400L
-    }
 
+    private val playerRepository by lazy { PlayerRepositoryImpl(context = applicationContext) }
```
**_Критическая_**

Слой представления (presentation-layer) не должен знать о data слое, в том числе об реализации репозиториев (PlayerRepositoryImpl). Для внедрения необходимых зависимостей тебе понадобится промежуточный класс, например, App.kt:

```
class App : Application() {
   companion object {
        internal lateinit var INSTANCE: App
            private set
   }
    
   lateinit var playerRepository: PlayerRepository
   
   override fun onCreate() {
      INSTANCE = this
      playerRepository = PlayerRepositoryImpl(context = applicationContext)
      ...
   }
}
```

И уже в AudioPlayerActivity.kt ты сможешь получить нужную зависимость следующим образом:
`private val playerRepository = App.INSTANCE.playerRepository `

Тем самым, у тебя не нарушаются границы между слоями.

IvanGusevE commented on app/src/main/java/com/example/playlistmaker/presentation/ui/AudioPlayerActivity.kt:
```diff
@@ -128,64 +119,4 @@ class AudioPlayerActivity : AppCompatActivity() {
-
-    }
-
-    private fun playbackControl() {
-        when (playerState) {
-            STATE_PLAYING -> {
-                pausePlayer()
-                handler.removeCallbacksAndMessages(null)
-            }
-
-            STATE_PREPARED, STATE_PAUSED -> {
-
-                startPlayer()
-                val startPlaying = System.currentTimeMillis()   //время начала отсчета
-                handler.post(createPlayingTimer(startPlaying))
-            }
-        }
-    }
```
Если перенесешь State в domain слой, то эту логику можно будет вернуть, т.к. у тебя появится возможность реагировать на события плеера из data слоя.

AntonSik replied:
> Если перенесешь State в domain слой, то эту логику можно будет вернуть, т.к. у тебя появится возможность реагировать на события плеера из data слоя.

Я окончательно запутался( выручайте, все сломалось. сделал новый пуш

IvanGusevE replied:
Привет, посмотрел твой коммит. Ты двигаешься в правильном направлении. Теперь твой репозиторий имеет метод `fun setOnChangePlayerListener(listener: (AudioPlayerState) -> Unit)`, при помощи которого твое активити сможет реагировать на смену состояния плеера:

```
playerRepository.setOnChangePlayerListener { state ->
   playBackControl(state)
}
```

В реализации репозитория тебе не нужно хранить текущее состояние, а наоборот, при помощи `audioPlayerListener` передавать его:

```
override fun pausePlayer() {
   mediaPlayer.pause()
   audioPlayerListener?.invoke(AudioPlayerState.PAUSED)
}
```

На самом деле, для еще большой простоты, можно удалить методы:
```
fun setOnPreparedCallback(callback: () -> Unit)
fun setOnCompletionCallBack(callback: () -> Unit) 
```

И оставить только `setOnChangePlayerListener`. Тогда твоя реализация репозитория будет следующей:

```
class PlayerRepositoryImpl(val context: Context) : PlayerRepository {
   
    private var audioPlayerListener: ((AudioPlayerState) -> Unit)? = null
    private val mediaPlayer = MediaPlayer()
    
    override fun startPlayer(callback: () -> Unit) {
        mediaPlayer.start()
        audioPlayerListener?.invoke(AudioPlayerState.PLAYING)
    }
    
    override fun pausePlayer(callback: () -> Unit) {
        mediaPlayer.pause()
        audioPlayerListener?.invoke(AudioPlayerState.PAUSED)
    }

    override fun preparePlayer(recordsUrl: String?) {
        mediaPlayer.setDataSource(recordsUrl)
        mediaPlayer.prepareAsync()
        mediaPlayer.setOnPreparedListener {
            audioPlayerListener?.invoke(AudioPlayerState.PREPARED)
        }
     }
     
     override fun setOnChangePlayerListener(listener: (AudioPlayerState) -> Unit) {
        audioPlayerListener = listener
    }
}
```

Также стоит удалить mediaplayer из активити. Чтобы получить `currentPosition` или вызвать `release()` нужно просто добавить два дополнительный метода в интерфейс твоего репозитория, например, `getCurrentPosition()` и `releasePlayer()`, и реализовать их в data слое.  

Но я написал часть реализации, тебе стоит его доработать. Возможно, для `releasePlayer` придется добавить дополнительное стотояние.

AntonSik replied:
Выручайте пожалуйста дорогой товарищ, ошибка такая случилась понимаю что связана с состоянием mediaPlayer, но где выстреливает не могу найти место(Ошибка случается если кликнуть  по треку из списка, затем вернуться на предыдущий экран результатов поиска а потом вновь кликнуть по любому элементу списка). Так же не понимаю почему не отрабатывает код AudioPlayerState.COMPLETED -> {
                    timePlaying.text = getString(R.string.zero)
                    playBtn.setImageResource(R.drawable.play_vector)
                    handler.removeCallbacksAndMessages(null)
            }  сделал коммит
![fatal error](https://github.com/AntonSik/PlaylistMaker/assets/121441647/6b4a9ddb-9859-4626-934b-273f8d673de9)

IvanGusevE replied:
Антон, привет! Проблема видимо в том, что `playerRepository` нужно пересоздавать каждый раз, когда пользователь переходит на экран выбранного трека, чтобы внутри повторно создать медиа плеер. Чтобы это исправить, в App тебе нужно:

Вместо этого:

```
lateinit var playerRepository : PlayerRepository

override fun onCreate() {
        super.onCreate()

        playerRepository = PlayerRepositoryImpl(context = applicationContext)
        ....
}
```

Написать следующее:

```
fun getPlayerRepository(): PlayerRepository = PlayerRepositoryImpl()

override fun onCreate() {
    super.onCreate()
    ....
}
```

Тогда при каждом вызове метода `getPlayerRepository()` у тебя будет создаваться новый репозиторий и проблема не должна возникать. Также тебе не нужно передавать контекст, т.к. в твоем репозитории он не используется. В `AudioPlayerActivity` останется только подправить:

`val playerRepository by lazy { App.INSTANCE.getPlayerRepository() }`

AntonSik replied:
Благодарю, это решило проблему, а вот с блоком кода как быть у меня есть mediaplayer.setOnCompletitionListener{ audioPlayerListener?.invoke(AudioPlayerState.COMPLETED) а код который вызывается по состоянию AudioPlayerState.COMPLETED не отрабатывает AudioPlayerState.COMPLETED -> {
timePlaying.text = getString(R.string.zero)
playBtn.setImageResource(R.drawable.play_vector)
handler.removeCallbacksAndMessages(null)
}

AntonSik replied:
я понял в чем косяк, косяк в том что в методе playBack блок Completed срабатывает только по нажатию на кнопку в которой он вызывается

AntonSik committed:
15th sprint was added:
- packages
- playerRepository interface and his implementation
- removed activity-files into ui package

AntonSik committed:
15th sprint third attempt

AntonSik committed:
15th sprint fourth attempt

AntonSik committed:
15th sprint 5th commit

IvanGusevE commented on app/src/main/java/com/example/playlistmaker/App.kt:
```diff
@@ -4,18 +4,27 @@ import android.app.Application
 
     var darkTheme = false
     private lateinit var sharedPrefs : SharedPreferences
+    fun getRepository() = PlayerRepositoryImpl()
```
Здесь лучше четко задать возвращаемое значение интерфейс `PlayerRepository`, чтобы не нарушать границы между слоями.

IvanGusevE commented on app/src/main/java/com/example/playlistmaker/domain/repository/PlayerRepository.kt:
```diff
@@ -0,0 +1,20 @@
+
+interface PlayerRepository {
+
+    fun startPlayer(callback: () -> Unit)
+
+    fun pausePlayer(callback: () -> Unit)
+
+    fun getDefault(callback: () -> Unit)
```
А зачем передается в каждом методе лямбда, которая не используется? В комментариях писал про то, что это нужно сделать для тех методов, которые устанавливают обработчик (setOnChangePlayerListener, setOnCompletionCallback).

AntonSik replied:
Аааааа я просто не понял зачем они нужны))), как-то эти колбэки весьма поверхностно прошли. Спасибо исправлю)

AntonSik committed:
15th sprint last commit
